### PR TITLE
Add/render layout support updates

### PR DESCRIPTION
**Pull request by tellthemachines**
Trac ticket: https://core.trac.wordpress.org/ticket/57584

Includes changes from:

* https://github.com/WordPress/gutenberg/pull/44600
* https://github.com/WordPress/gutenberg/pull/45330 
which change the layout render filter to add layout classes to the inner block wrapper, instead of the outer, and
* https://github.com/WordPress/gutenberg/pull/45364
* https://github.com/WordPress/gutenberg/pull/46139
which change the layout render filter to allow inner blocks to receive layout classes and styles. This builds on the former changeset, which is why I thought it simpler to backport both pieces together. 

I also added some tests for `wp_render_layout_support_flag` (from https://github.com/WordPress/gutenberg/pull/47719) and updated some of the server block fixtures because the change to add classnames to the inner block wrapper caused the order in which layout classes are added to switch.

To test the child layout changes, add the following markup in the editor, save and check that child block 1 occupies all the leftover width of its container:

```
<!-- wp:group {"layout":{"type":"flex","flexWrap":"nowrap","verticalAlignment":"stretch"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"layout":{"selfStretch":"fill"}},"backgroundColor":"luminous-vivid-amber"} -->
<p class="has-luminous-vivid-amber-background-color has-background">1 </p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"backgroundColor":"luminous-vivid-amber"} -->
<p class="has-luminous-vivid-amber-background-color has-background">2</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->
```

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/block-supports/layout.php [hunk 1]
```diff
@@ -308,6 +308,26 @@ function wp_get_layout_style( $selector, $layout, $has_block_gap_support = false
...
+function wp_get_classnames_from_last_tag( $html ) {
+	$tags            = new WP_HTML_Tag_Processor( $html );
+	$last_classnames = '';
+
+	while ( $tags->next_tag() ) {
+		$last_classnames = $tags->get_attribute( 'class' );
+	}
+
+	return (string) $last_classnames;
+}
```
Comment (Mamaduka) [inline src/wp-includes/block-supports/layout.php]: Suggestion: I only see a single use of this method. Maybe we should avoid introducing a public method and use `WP_HTML_Tag_Processor` directly.

 The functions are always public in PHP, and we can only deprecate them after the release.

**Review by ntsekouras — COMMENTED** [review]
Besides @Mamaduka 's comment about the new function vs inlining, I tested and this looks good to me. Thanks!

I've also tested the `layout` changes on the packages update PR and also tested the `stack` setting(having the updated packages) with this markup:

```
<!-- wp:group {"style":{"dimensions":{"minHeight":"400px"}},"layout":{"type":"flex","orientation":"vertical"}} -->
<div class="wp-block-group" style="min-height:400px"><!-- wp:heading {"style":{"layout":{"selfStretch":"fill","flexSize":null}},"backgroundColor":"light-green-cyan"} -->
<h2 class="wp-block-heading has-light-green-cyan-background-color has-background">1</h2>
<!-- /wp:heading -->

<!-- wp:paragraph {"style":{"layout":{"selfStretch":"fixed","flexSize":"123px"}},"backgroundColor":"luminous-vivid-orange"} -->
<p class="has-luminous-vivid-orange-background-color has-background">2</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->
```

File: tests/phpunit/tests/block-supports/layout.php [hunk 2]
```diff
@@ -173,4 +173,88 @@ public function test_outer_container_not_restored_for_aligned_image_block_with_t
...
+	 * @dataProvider data_layout_support_flag_renders_classnames_on_wrapper
+	 *
+	 * @covers ::wp_render_layout_support_flag
+	 *
+	 * @param array  $args            Dataset to test.
+	 * @param string $expected_output The expected output.
+	 */
+	public function test_layout_support_flag_renders_classnames_on_wrapper( $args, $expected_output ) {
+		$actual_output = wp_render_layout_support_flag( $args['block_content'], $args['block'] );
+		$this->assertEquals( $expected_output, $actual_output );
```
Comment (dream-encode) [inline tests/phpunit/tests/block-supports/layout.php]:
```suggestion
		$this->assertSame( $expected_output, $actual_output );
```

File: src/wp-includes/block-supports/layout.php [hunk 3]
```diff
@@ -321,10 +341,56 @@ function wp_get_layout_style( $selector, $layout, $has_block_gap_support = false
...
 
-	if ( ! $support_layout ) {
+	if ( ! $support_layout
+	&& ! $has_child_layout ) {
 		return $block_content;
 	}
+	$outer_class_names = array();
+
+	if ( $has_child_layout && ( 'fixed' === $block['attrs']['style']['layout']['selfStretch'] || 'fill' === $block['attrs']['style']['layout']['selfStretch'] ) ) {
+
```
Comment (dream-encode) [inline src/wp-includes/block-supports/layout.php]:
```suggestion
```

File: src/wp-includes/block-supports/layout.php [hunk 4]
```diff
@@ -321,10 +341,56 @@ function wp_get_layout_style( $selector, $layout, $has_block_gap_support = false
 function wp_render_layout_support_flag( $block_content, $block ) {
 	$block_type     = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
 	$support_layout = block_has_support( $block_type, array( '__experimentalLayout' ), false );
+	$has_child_layout = isset( $block['attrs']['style']['layout']['selfStretch'] );
 
-	if ( ! $support_layout ) {
+	if ( ! $support_layout
+	&& ! $has_child_layout ) {
```
Comment (dream-encode) [inline src/wp-includes/block-supports/layout.php]:
```suggestion
	if ( ! $support_layout && ! $has_child_layout ) {
```

File: src/wp-includes/block-supports/layout.php [hunk 5]
```diff
@@ -321,10 +341,56 @@ function wp_get_layout_style( $selector, $layout, $has_block_gap_support = false
...
+		wp_style_engine_get_stylesheet_from_css_rules(
+			$child_layout_styles,
+			array(
+				'context'  => 'block-supports',
+				'prettify' => false,
+			)
+		);
+
+		$outer_class_names[] = $container_content_class;
+
```
Comment (dream-encode) [inline src/wp-includes/block-supports/layout.php]:
```suggestion
```

File: src/wp-includes/block-supports/layout.php [hunk 6]
```diff
@@ -432,18 +497,39 @@ function wp_render_layout_support_flag( $block_content, $block ) {
...
+		foreach ( $outer_class_names as $outer_class_name ) {
+			$content_with_outer_classnames->add_class( $outer_class_name );
+		}
+
+		$content_with_outer_classnames = (string) $content_with_outer_classnames;
+	}
+
+	/**
+	* The first chunk of innerContent contains the block markup up until the inner blocks start.
+	* We want to target the opening tag of the inner blocks wrapper, which is the last tag in that chunk.
```
Comment (dream-encode) [inline src/wp-includes/block-supports/layout.php]:
```suggestion
	* This targets the opening tag of the inner blocks wrapper, which is the last tag in that chunk.
```

File: src/wp-includes/block-supports/layout.php [hunk 7]
```diff
@@ -308,6 +308,26 @@ function wp_get_layout_style( $selector, $layout, $has_block_gap_support = false
...
+/**
+ * Gets classname from last tag in a string of HTML.
+ *
+ * @since 6.2.0
+ * @access private
+ *
+ * @param string $html markup to be processed.
+ * @return string String of inner wrapper classnames.
+ */
+function wp_get_classnames_from_last_tag( $html ) {
```
Comment (dream-encode) [inline src/wp-includes/block-supports/layout.php]: Obligatory "Naming things is hard", but I wonder if more context in this function name is needed.

Why?

"Tag" could mean one of at least two things when it comes to WordPress:  a taxonomy type or a HTML element.

@hellofromtonya @felixarntz Any thoughts on the naming here?

**Review by dream-encode — CHANGES_REQUESTED** [review]

File: src/wp-includes/block-supports/layout.php (hunk 1 above)
Comment (felixarntz) [reply src/wp-includes/block-supports/layout.php]: @Mamaduka Agreed. Can we instead inline this logic in the `wp_render_layout_support_flag()` function please? This means we can avoid introducing this function.

If there is eventually another change that needs something like this, we can always introduce a new function, but right now it seems unnecessary.

File: src/wp-includes/block-supports/layout.php (hunk 7 above)
Comment (felixarntz) [reply src/wp-includes/block-supports/layout.php]: I think the name is okay (WordPress core already uses the word "tag" for both of the two meanings in various function names), but more importantly I think this conversation shouldn't really be relevant given @Mamaduka's feedback that we should not need to introduce this function in the first place.

File: src/wp-includes/block-supports/layout.php [hunk 8]
```diff
@@ -432,18 +497,39 @@ function wp_render_layout_support_flag( $block_content, $block ) {
...
+		}
+
+		$content_with_outer_classnames = (string) $content_with_outer_classnames;
+	}
+
+	/**
+	* The first chunk of innerContent contains the block markup up until the inner blocks start.
+	* We want to target the opening tag of the inner blocks wrapper, which is the last tag in that chunk.
+	*/
+	$inner_content_classnames = isset( $block['innerContent'][0] ) && 'string' === gettype( $block['innerContent'][0] ) ? wp_get_classnames_from_last_tag( $block['innerContent'][0] ) : '';
```
Comment (felixarntz) [inline src/wp-includes/block-supports/layout.php]: See above, can we inline that logic here in a dedicated `if ( isset( $block['innerContent'][0] ) && 'string' === gettype( $block['innerContent'][0] ) )` block?

**Review by felixarntz — CHANGES_REQUESTED** [review]
@tellthemachines Looks almost good to go from my perspective, though I agree with @Mamaduka that we should not unnecessarily introduce a new function here when its logic is rather specific and only needed in a single place.

**Comment by tellthemachines** [comment]
Thanks for the feedback folks! I've made all the requested changes, so this should be ready for another round.

**Comment by felixarntz** [comment]
@tellthemachines I tried to fix the merge conflict in the fixtures with latest `trunk` here, hopefully I did it right 🤞 

If there are test failures, I'll check and can hopefully fix it. Apologies, not really familiar with the fixtures.

**Review by dream-encode — APPROVED** [review]
Seems all feedback was addressed.  Looks good to merge.

**Review by felixarntz — APPROVED** [review]
Thanks @tellthemachines!

**Comment by felixarntz** [comment]
Noting that the test failures here are unrelated and intermittent, fixed via https://core.trac.wordpress.org/changeset/55280.

**Comment by felixarntz** [comment]
Committed in https://core.trac.wordpress.org/changeset/55282.

**Pull request closed without merging**